### PR TITLE
Replace the scale-blind element-pool solve with a staged, verified root-find

### PR DESCRIPTION
**Pull request by bjmorgan**
## Summary

For element-concentration constraints (`element_pools`), the coupled chemical-potential solve minimised a convex grand potential with scipy's `trust-exact` method, which terminates on an absolute gradient norm (default `gtol = 1e-4`). The gradient of that potential is measured in defects per unit cell, and realistic defect concentrations span roughly 1e-2 to 1e-20 per cell, so for any dilute target the optimiser accepted its starting point `mu = 0` unchanged and returned the unconstrained Boltzmann populations with the constraints silently ignored. Even at O(1) targets, convergence was limited to about 1e-4 absolute.

## The staged solve

The minimisation is replaced by a staged solve in a new module, `py_sc_fermi/element_pools.py`, with each stage engaged only when the previous stage's independently measured relative deviation misses tolerance:

1. A root-find (scipy `hybr`) on the dimensionless system `content_X(mu) / target_X - 1 = 0`, with the per-element-scaled Hessian -- the full state-ensemble covariance, including the empty state -- as analytic Jacobian, seeded by a diagonal Newton step of the log-content system from `mu = 0`. Scaling each equation by its own target keeps every residual O(1) however dilute the targets and however many orders of magnitude separate them; this settles the common case.
2. The same method on the log residual, whose full Jacobian resolves shared-species coupling (one species carrying most of several pools' content) that defeats the diagonal seed.
3. A derivative-free bracketing sweep: cyclic exact line solves of the grand potential along each element's own chemical potential, the collective all-ones direction, and a regularised Newton direction, each root bracketed by sign and polished with one-dimensional Newton on the exact line slope. It needs only residual signs, so saturation plateaux and underflowed Boltzmann weights cannot stall it, and it is the entry point for zero and negative net-content targets, which have no scaled residual or log seed.

The previous dedicated single-element `brentq` path is subsumed by the same code.

## Verification independent of the solver

No stage's own verdict is trusted: after any nominally successful solve, every element's achieved content is recomputed and checked against its target, and a deviation beyond 1e-6 relative raises an error naming the element, its target, and the achieved content. The check fails closed on NaN. The guarantee is asymmetric -- a returned solve meets tolerance, and a feasible solve the stages happen to miss raises loudly rather than returning a wrong number.

All element-pool constraint failures -- fixed concentrations inconsistent with a target, targets infeasible in either direction, and unmet targets -- are raised as `ElementPoolError` (a `ValueError` subclass), which propagates undisguised through `get_sc_fermi` rather than being relabelled as a Fermi-window failure.

## Constraint semantics

- An element whose target is fully committed by fixed concentrations is handled as the `lambda_X -> 0` limit: variable states with positive stoichiometry in that element get concentration zero and the element drops out of the solve. Remainders within rounding distance of the committed total are treated as exactly met, so float summation of the fixed contributions (`0.1 + 0.2 != 0.3`) cannot raise a spurious inconsistency error.
- Pools with a negative-stoichiometry variable state can shed content, so zero and negative net-content targets over mixed-sign stoichiometries are balance conditions: an off-stoichiometry scan crosses exact stoichiometry continuously. A feasibility bound on the minimum achievable content mirrors the existing maximum-content bound. Small net targets over near-saturated sites resolve down to about 1e-11 per cell; below that, catastrophic cancellation between the opposing populations is fundamental and the solve fails loud.

## Structure

The solver operates on a plain per-group representation (`GroupData`: free-site count, log Boltzmann weights, stoichiometry vectors), independent of `DefectSystem`. `DefectSystem._global_defect_concs` now reads as build groups -> solve pools -> deliver concentrations, with the pool disposition and solve in a `_solve_element_pools` helper, and the per-state triples carried by a `_VariableState` NamedTuple.

## Tests

New convergence tests cover coupled dilute targets (~1e-17 per cell), coupled targets twelve orders of magnitude apart, O(1) targets reached from deeply dilute unconstrained populations, exactly underflowed Boltzmann weights (2 eV at 30 K), saturated species, charged dopants solved through `get_sc_fermi`, and small net targets at half-saturation. Edge-case tests cover exhausted pools, zero and negative balance targets, infeasibility in both directions, the rounding clamp, and the misreported-convergence guard. A new `tests/test_element_pools.py` unit-tests the solver primitives directly on hand-built `GroupData`. The existing coupled-pool test is tightened from `places=3` absolute to 1e-8 relative.

### CODE REVIEW

File: py_sc_fermi/defect_system.py [hunk 1]
```diff
@@ -385,44 +386,33 @@ def _remaining_element_targets(
...
-                    f"target {target:.3e}. Your constraints are mutually "
-                    "inconsistent."
-                )
+            # Float summation of fixed contributions does not land exactly
+            # on a target they are meant to meet (0.1 + 0.2 != 0.3): treat
+            # a remainder within rounding distance of the committed total
+            # as exactly zero. Genuinely negative remainders are validated
+            # by the caller, which knows whether the pool can shed content.
+            if abs(rem) <= 1e-12 * committed:
+                rem = 0.0
```
[the remark this answers is not in the document]
Comment (bjmorgan) [reply py_sc_fermi/defect_system.py]: Fixed in 425d010: the clamp scales by `abs(committed)` so it triggers when fixed negative-stoichiometry states make the committed total negative. Covered by `test_exhausted_pool_tolerates_rounding_with_negative_committed` (committed -0.3, target -(0.1 + 0.2)).

**Comment by bjmorgan** [comment]
Merged by fast-forwarding dev to the branch head (8eff0b7), rather than through the GitHub merge button: dev had not diverged since the branch point, so the 22 commits land on dev exactly as reviewed, with their original SHAs and no merge or squash commit. The branch history was kept deliberately -- each commit is a coherent, test-green change whose message records the reasoning behind the numerical design decisions, which squashing would have discarded.
